### PR TITLE
Property Tests: make_reward_set proptest and documentation

### PR DESCRIPTION
**Pull request by aaronb-stacks**
This PR adds some docs for how we should handle property testing in this repo going forward, and it includes an example property test for the `make_reward_set` function.

### CODE REVIEW

File: docs/property-testing.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    Ok(())
+}
+```
+
+This test essentially just takes the raw PoX entries, computes the PoX stacking threshold, and then feeds that data into `make_reward_set`. Afterwards, it checks that each PoX entry has the expected number of slots (i.e., the sum of all of that address's entries floor-divided by the threshold). The one "proptest hack" in this function is the way total liquid ustx is computed. Rather than allowing total liquid ustx to be a free argument and then just "prop assuming" that it is greater than the sum of the reward set entries, we make total liquid ustx a derived variable. The reason to do this is that it makes input generation easier if it doesn't have to worry too much about generating invalid data.
+
+Now, the part of the test that becomes more complex (and proptest-specific) is the actual input generation. We need strategies for generating the inputs to that function. For `PoxConstants`, we're really only interesting in the number of slots, so we can just generate uints and construct `PoxConstants` from that. Similarly, `unstacked_ustx` is just a `u128`. However, we do have to write a strategy for the reward set entries themselves.
```
Comment (brice-stacks) [inline docs/property-testing.md]:
```suggestion
Now, the part of the test that becomes more complex (and proptest-specific) is the actual input generation. We need strategies for generating the inputs to that function. For `PoxConstants`, we're really only interested in the number of slots, so we can just generate uints and construct `PoxConstants` from that. Similarly, `unstacked_ustx` is just a `u128`. However, we do have to write a strategy for the reward set entries themselves.
```

File: docs/property-testing.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        check_make_reward_set(pox_settings, addrs.as_slice(), unstacked_ustx)?;
+    }
+```
+
+This technique allows to be sure that proptest generates a lot of cases where there are multiple entries for the same reward address. Unfortunately, this kind of thing tends to be more art than science, which means that PR authors and reviewers will need to be careful about the input strategies for property tests (this should also be aided by the CI task for PRs). This is one of the reasons that property tests can't totally supplant unit tests. However, a lot of the work of property tests helps with writing unit tests: many unit tests can be essentially fixed inputs to the property test.
+
+## Reusing Strategies 
+
+Writing new input strategies may be the most tedious part of writing property tests, so it is worthwhile figuring out if the input you are looking for (or maybe a component of the input you're looking for) already has a strategy in the codebase. If you search for functions that return `impl Strategy<Value = ?>` in the codebase, you should find the set of functions that have already been written.
```
Comment (brice-stacks) [inline docs/property-testing.md:206]: Should we have a convention of defining these all in some common place, for discoverability?

File: docs/property-testing.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+## Continuous Integration
+
+By default, we'll get some CI integration from `proptest` automatically: the new property tests will run with 250 randomly generated inputs on every execution of the unit test job in CI. This is great. However, we want some additional support for executing *new* property tests extra amounts before PRs merge.
+
+The environment variable `PROPTEST_CASES` can be set to a higher number (e.g., `PROPTEST_CASES=2500`) to explore more test cases before declaring success. From the CI, what we want is a job which:
+
+1. Executes once a PR has been approved.
+2. Discovers the set of new tests (this is probably easiest to achieve by running `cargo nextest list` on the source and target branches and then diffing the outputs).
+3. Executes only the new tests with the environment variable `PROPTEST_CASES` set to 2500.
```
Comment (brice-stacks) [inline docs/property-testing.md:218]: Great idea!

File: docs/property-testing.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            let mut new_entry = addrs[to_dup_ix % addrs.len()].clone();
+            new_entry.amount_stacked = duplicated_amount;
+            addrs.push(new_entry);
+        }
+
+        check_make_reward_set(pox_settings, addrs.as_slice(), unstacked_ustx)?;
+    }
+```
+
+This technique allows to be sure that proptest generates a lot of cases where there are multiple entries for the same reward address. Unfortunately, this kind of thing tends to be more art than science, which means that PR authors and reviewers will need to be careful about the input strategies for property tests (this should also be aided by the CI task for PRs). This is one of the reasons that property tests can't totally supplant unit tests. However, a lot of the work of property tests helps with writing unit tests: many unit tests can be essentially fixed inputs to the property test.
```
Comment (brice-stacks) [inline docs/property-testing.md:202]: I like the idea of using a common function which is called in the property test(s) and also called with some known edge cases in standard unit test style. Can we include that in the example here?

File: docs/property-testing.md (hunk 2 above)
Comment (aaronb-stacks) [reply docs/property-testing.md:206]: Yep -- moved the functions and added a note in def7501

File: docs/property-testing.md (hunk 4 above)
Comment (aaronb-stacks) [reply docs/property-testing.md:202]: Yep, included in [def7501](https://github.com/stacks-network/stacks-core/pull/6640/commits/def750187848210742009118da73158913889b4e)

**Review by brice-stacks — APPROVED** [review]
LGTM, but there are some clippy issues preventing the tests from running.

File: stackslib/src/chainstate/stacks/tests/reward_set.rs [hunk 5]
```diff
@@ -0,0 +1,237 @@
+use std::collections::HashMap;
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/tests/reward_set.rs:16]: Missing copyright statement

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/tests/reward_set.rs:16]: (we should really have a CI check for this)

**Review by jcnelson — DISMISSED** [review]
LGTM; just please add copyright comment blocks 🙏

**Review by jcnelson — APPROVED** [review]

File: docs/property-testing.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+## Examples
+
+There are a couple examples of proptest in action in our codebase. The first is a set of property tests for new clarity functions in Epoch 3.3 (Clarity4). These tests generate clarity code (as strings) for property tests that assert the new clarity functions behave as expected ([vm::tests::post_conditions](clarity/src/vm/tests/post_conditions.rs#L1761)). The second is a property test for `make_reward_set`, which is used to translate data pulled from the `pox` contracts into the actual reward set ([chainstate::stacks::tests::reward_set](stackslib/src/chainstate/stacks/tests/reward_set.rs)).
```
Comment (federico-stacks) [inline docs/property-testing.md]: nit: It might be helpful to use a permalink or similar for these links, so it won’t break in the future and the documentation remains consistent.

Comment (aaronb-stacks) [reply docs/property-testing.md]: Good idea -- updated
